File: tests/mjorbit/reference/orbit/elements.py
"""Keplerian orbital element conversions.

Units: km, km/s, rad.
Ported from OrbitX elements/conversions.py (Vallado Algorithm 10).
"""

from __future__ import annotations

import numpy as np

from mjorbit.constants import GM_EARTH


def keplerian_to_cartesian(
    a: float,
    e: float,
    inc: float,
    raan: float,
    argp: float,
    nu: float,
) -> tuple[np.ndarray, np.ndarray]:
    """Convert Keplerian elements to ECI Cartesian state.

    Args:
        a: semi-major axis, km
        e: eccentricity
        inc: inclination, rad
        raan: right ascension of ascending node, rad
        argp: argument of perigee, rad
        nu: true anomaly, rad

    Returns:
        (R_eci, V_eci) each shape (3,) in km and km/s
    """
    p = a * (1.0 - e**2)
    r_pqw = p / (1.0 + e * np.cos(nu))

    # Position and velocity in perifocal frame
    cos_nu, sin_nu = np.cos(nu), np.sin(nu)
    R_pf = r_pqw * np.array([cos_nu, sin_nu, 0.0])
    V_pf = np.sqrt(GM_EARTH / p) * np.array([-sin_nu, e + cos_nu, 0.0])

    # Rotation matrices from perifocal to ECI
    # R3(-RAAN) @ R1(-inc) @ R3(-argp)
    R = _rot_pf_to_eci(raan, inc, argp)
    return R @ R_pf, R @ V_pf


def cartesian_to_keplerian(
    R_eci: np.ndarray, V_eci: np.ndarray
) -> dict[str, float]:
    """Convert ECI Cartesian state to Keplerian elements.

    Returns dict with keys: a, e, inc, raan, argp, nu (rad, km).
    """
    r = np.linalg.norm(R_eci)
    v = np.linalg.norm(V_eci)
    mu = GM_EARTH

    h = np.cross(R_eci, V_eci)
    h_mag = np.linalg.norm(h)

    n = np.cross(np.array([0.0, 0.0, 1.0]), h)
    n_mag = np.linalg.norm(n)

    e_vec = ((v**2 - mu / r) * R_eci - np.dot(R_eci, V_eci) * V_eci) / mu
    e = np.linalg.norm(e_vec)

    energy = v**2 / 2.0 - mu / r
    a = -mu / (2.0 * energy)

    inc = np.arccos(h[2] / h_mag)

    raan = np.arctan2(n[1], n[0]) if n_mag > 1e-12 else 0.0

    argp = np.arccos(np.clip(np.dot(n, e_vec) / (n_mag * e + 1e-30), -1.0, 1.0))
    if e_vec[2] < 0.0:
        argp = 2.0 * np.pi - argp

    nu = np.arccos(np.clip(np.dot(e_vec, R_eci) / (e * r + 1e-30), -1.0, 1.0))
    if np.dot(R_eci, V_eci) < 0.0:
        nu = 2.0 * np.pi - nu

    return dict(a=a, e=e, inc=inc, raan=raan, argp=argp, nu=nu)


def _rot_pf_to_eci(raan: float, inc: float, argp: float) -> np.ndarray:
    """3x3 rotation from perifocal to ECI frame."""
    c_O, s_O = np.cos(raan), np.sin(raan)
    c_i, s_i = np.cos(inc), np.sin(inc)
    c_w, s_w = np.cos(argp), np.sin(argp)

    R = np.array([
        [c_O * c_w - s_O * s_w * c_i,  -c_O * s_w - s_O * c_w * c_i,   s_O * s_i],
        [s_O * c_w + c_O * s_w * c_i,  -s_O * s_w + c_O * c_w * c_i,  -c_O * s_i],
        [s_w * s_i,                      c_w * s_i,                       c_i      ],
    ])
    return R

File: src/cpp/src/elements.cc
#include "mjorbit/elements.h"

#include <cmath>

#include "mjorbit/math_utils.h"

namespace mjorbit {

void rot_pf_to_eci(double raan, double inc, double argp, double out_R[9]) {
  const double c_O = std::cos(raan);
  const double s_O = std::sin(raan);
  const double c_i = std::cos(inc);
  const double s_i = std::sin(inc);
  const double c_w = std::cos(argp);
  const double s_w = std::sin(argp);

  out_R[0] = c_O * c_w - s_O * s_w * c_i;
  out_R[1] = -c_O * s_w - s_O * c_w * c_i;
  out_R[2] = s_O * s_i;
  out_R[3] = s_O * c_w + c_O * s_w * c_i;
  out_R[4] = -s_O * s_w + c_O * c_w * c_i;
  out_R[5] = -c_O * s_i;
  out_R[6] = s_w * s_i;
  out_R[7] = c_w * s_i;
  out_R[8] = c_i;
}

void keplerian_to_cartesian(
    double a,
    double e,
    double inc,
    double raan,
    double argp,
    double nu,
    double out_R_eci[3],
    double out_V_eci[3],
    double gm) {
  const double p = a * (1.0 - e * e);
  const double r_pqw = p / (1.0 + e * std::cos(nu));
  const double cos_nu = std::cos(nu);
  const double sin_nu = std::sin(nu);

  const double R_pf[3] = {r_pqw * cos_nu, r_pqw * sin_nu, 0.0};
  const double sqrt_mu_over_p = std::sqrt(gm / p);
  const double V_pf[3] = {-sqrt_mu_over_p * sin_nu, sqrt_mu_over_p * (e + cos_nu), 0.0};

  double rot[9];
  rot_pf_to_eci(raan, inc, argp, rot);
  detail::mat3_mul_vec(rot, R_pf, out_R_eci);
  detail::mat3_mul_vec(rot, V_pf, out_V_eci);
}

void cartesian_to_keplerian(
    const double R_eci[3],
    const double V_eci[3],
    double* out_a,
    double* out_e,
    double* out_inc,
    double* out_raan,
    double* out_argp,
    double* out_nu,
    double gm) {
  const double r = detail::norm3(R_eci);
  const double v = detail::norm3(V_eci);

  double h[3];
  detail::cross3(R_eci, V_eci, h);
  const double h_mag = detail::norm3(h);

  const double k_hat[3] = {0.0, 0.0, 1.0};
  double n[3];
  detail::cross3(k_hat, h, n);
  const double n_mag = detail::norm3(n);

  const double rv_dot = detail::dot3(R_eci, V_eci);
  const double coeff = v * v - gm / r;
  double e_vec[3];
  for (int i = 0; i < 3; ++i) {
    e_vec[i] = (coeff * R_eci[i] - rv_dot * V_eci[i]) / gm;
  }
  const double e = detail::norm3(e_vec);

  const double energy = 0.5 * v * v - gm / r;
  const double a = -gm / (2.0 * energy);
  const double inc = std::acos(h[2] / h_mag);
  const double raan = (n_mag > 1e-12) ? std::atan2(n[1], n[0]) : 0.0;

  double argp = std::acos(detail::clamp(detail::dot3(n, e_vec) / (n_mag * e + 1e-30), -1.0, 1.0));
  if (e_vec[2] < 0.0) {
    argp = 2.0 * detail::kPi - argp;
  }

  double nu = std::acos(detail::clamp(detail::dot3(e_vec, R_eci) / (e * r + 1e-30), -1.0, 1.0));
  if (rv_dot < 0.0) {
    nu = 2.0 * detail::kPi - nu;
  }

  if (out_a) {
    *out_a = a;
  }
  if (out_e) {
    *out_e = e;
  }
  if (out_inc) {
    *out_inc = inc;
  }
  if (out_raan) {
    *out_raan = raan;
  }
  if (out_argp) {
    *out_argp = argp;
  }
  if (out_nu) {
    *out_nu = nu;
  }
}

}  // namespace mjorbit
